src/logging.rs

```diff
@@ -18,38 +14,105 @@ thread_local! {
     static CONTEXT: RefCell<Vec<*const fmt::Display>> = RefCell::new(Vec::new());
 }
 
-pub fn formatted_builder() -> env_logger::Builder {
-    let start_time = clock::now();
-    let mut builder = env_logger::Builder::new();
-    builder.format(move |fmt, record| {
-        CONTEXT.with(move |ctxt| {
-            let level = match record.level() {
-                Level::Trace => "TRCE",
-                Level::Debug => "DBUG",
-                Level::Info => "INFO",
-                Level::Warn => "WARN",
-                Level::Error => "ERR!",
+pub mod trace {
+    extern crate tokio_trace_log;
+    use super::{clock, Context as LegacyContext, CONTEXT as LEGACY_CONTEXT};
+    use std::{env, fmt, sync::Arc};
+    pub use tokio_trace::*;
+    pub use tokio_trace_fmt::*;
+
+    pub type SubscriberBuilder = Builder<
+        default::NewRecorder,
+        Box<
+            Fn(&Context<default::NewRecorder>, &mut fmt::Write, &Event) -> fmt::Result
+                + Send
+                + Sync
+                + 'static,
+        >,
+        filter::EnvFilter,
+    >;
+
+    pub fn init() {
+        try_init().expect("initializing tokio-trace failed");
+    }
+
+    pub fn try_init() -> Result<(), &'static str> {
+        let env = env::var(super::ENV_LOG).unwrap_or_default();
+        try_init_with_filter(env)
+    }
+
+    pub fn try_init_with_filter<F: AsRef<str>>(filter: F) -> Result<(), &'static str> {
+        let dispatch = Dispatch::new(
+            subscriber_builder()
+                .with_filter(filter::EnvFilter::from(filter))
+                .finish(),
+        );
+        dispatcher::set_global_default(dispatch)
+            .map_err(|_| "failed to set global default dispatcher")?;
+        let logger = tokio_trace_log::LogTracer::with_filter(log::LevelFilter::max());
+        log::set_boxed_logger(Box::new(logger)).map_err(|_| "failed to set global logger")?;
+        log::set_max_level(log::LevelFilter::max());
```

Comment:
Is `set_max_level` still needed with the `with_filter(log::LevelFilter::max())` on :52?

Reply:
Yes, they are unrelatedd. `set_max_level` configures the _global_ max level for the `log` crate; `with_filter` configures a filter for the `LogTracer` adapter (which is not necessary as the trace subscriber manages filtering).